scanpy/external/_tools/_annotator.py

```diff
@@ -0,0 +1,116 @@
+from anndata import AnnData
+import pandas as pd
+
+
+def annotator(adata: AnnData, markers: pd.DataFrame, num_genes: int,
+              return_nonzero_annotations: bool=True, p_threshold: float=0.05,
+              p_value_fun: str="binom", z_threshold: float=1.,
+              scoring: str="exp_ratio", normalize: bool=False):
+    """
+    Annotator marks the data with cell type annotations based on marker genes.
+    Over-expressed genes are selected with the  Mann-Whitney U tests and cell
+    types are assigned with the hypergeometric test. This function first selects
+    genes from gene expression data with the Mann Whitney U test, then annotate
+    them with the hypergeometric test, and finally filter out cell types that
+    have zero scores for all cells. The results are scores that tell how
+    probable is each cell type for each cell.
+
+    Parameters
+    ----------
+    adata : AnnData
+        Tabular data with gene expressions.
+    markers : pd.DataFrame
+        The data-frame with marker genes and cell types. Data-frame has two
+        columns Gene and Cell Type first holds gene names or ID and second cell
+        type for this gene. Gene names must be written in the same format than
+        genes in `adata`.
+    num_genes : int
```

Comment:
The current way is a bit vague because you have to know this value.

If no value is provided, use the number of columns in the gene expression matrix. Perhaps issue a warning or message when this happens.

Reply:
Done